File: solutions/cpp/0039.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> curr;
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0, curr, ans);
        return ans;
    }

private:
    void dfs(vector<int>& candidates, int target, int s, vector<int>& curr, vector<vector<int>>& ans) {
        if (target < 0) return;
        if (target == 0) {
            ans.push_back(curr);
            return;
        }

        for (int i = s; i < candidates.size(); i++) {
            curr.push_back(candidates[i]);
            dfs(candidates, target - candidates[i], i, curr, ans);
            curr.pop_back();
        }
    }
};

File: solutions/python3/0039.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        candidates.sort()
        self.dfs(candidates, target, 0, [], ans)
        return ans

    def dfs(self, candidates: List[int], target: int, s: int, curr: List[int], ans: List[List[int]]) -> None:
        if target < 0:
            return
        if target == 0:
            ans.append(curr)
            return

        for i in range(s, len(candidates)):
            self.dfs(candidates, target - candidates[i], i, curr + [candidates[i]], ans)